### PR TITLE
[ShapesSpec] Extract spec-binding helpers into shared fx modules.

### PR DESCRIPTION
**Pull request by laithsakka**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* #187010
* #187602
* #186751
* __->__ #187160

Pure code movement to prepare for make_fx dynamic-shape support, with no
behavior change for torch.export.

1. The spec->args binding helpers (_walk_spec / _bind_spec_to_args) move out of
torch/_dynamo/functional_export.py into a new fx-layer module
torch/fx/experimental/_spec_binding.py. 

2.  ShapeEnv wiring helpers (_symbolic_context_from_shapes_spec, 
_wire_tensor_spec_dims, _wire_spec_slot,
_wire_spec_assumptions, _finalize_spec_wiring) move out of
torch/_dynamo/variables/builder.py into torch/fx/experimental/symbolic_shapes.py.
_coerce_to_shapes_spec is added to torch/fx/experimental/dynamic_spec.py so the
dynamo and fx entry points share one normalization path.

3.  Call sites (output_graph.py, eval_frame.py, functional_export.py) and the relocated unit
tests are updated to the new import locations.

_emit_pending_bool is renamed to _emit_pending_spec_assumption
_wire_tensor_spec_dims introduce (iterate on tensor and call _wire_spec_slot_
cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @jataylo @azahed98

### CODE REVIEW

File: torch/_dynamo/variables/builder.py [hunk 1]
```diff
@@ -4965,14 +4770,7 @@ def _wrap_to_fake_tensor_and_record_impl(
                 )
             )
         if tensor_spec is not None:
-            for dim_i in range(fake_e.dim()):
```
Comment (laithsakka) [inline torch/_dynamo/variables/builder.py:4996]: _wire_tensor_spec_dims is a function that includes this block

**Comment by jansel** [comment]
@claude confirm that this is just code movement and doesn't change semantics

**Review by jansel — APPROVED** [review]
